geomesa-memory/geomesa-cqengine-datastore/src/main/scala/org/locationtech/geomesa/memory/cqengine/datastore/GeoCQEngineFeatureStore.scala

```diff
@@ -57,17 +39,62 @@ class GeoCQEngineFeatureStore(engine: GeoCQEngine, entry: ContentEntry, query: Q
   override def getCountInternal(query: Query): Int =
     SelfClosingIterator(engine.query(query.getFilter)).length
 
-  override def getReaderInternal(query: Query): FeatureReader[SimpleFeatureType, SimpleFeature] = {
-    import scala.collection.JavaConversions._
-    new DelegateFeatureReader(engine.sft, new DelegateFeatureIterator(engine.query(query.getFilter)))
-  }
+  override def getReaderInternal(query: Query): FeatureReader[SimpleFeatureType, SimpleFeature] =
+    new DelegateFeatureReader(engine.sft, new DelegateFeatureIterator(engine.query(query.getFilter).asJava))
+
+  override def getQueryCapabilities: QueryCapabilities = GeoCQEngineQueryCapabilities
+
+  override def canEvent: Boolean = true
+}
+
+object GeoCQEngineFeatureStore {
 
   object GeoCQEngineQueryCapabilities extends QueryCapabilities {
     override def isUseProvidedFIDSupported = true
   }
 
-  override def getQueryCapabilities = GeoCQEngineQueryCapabilities
-
-  override def canEvent: Boolean = true
-
+  class GeoCQEngineFeatureWriter(engine: GeoCQEngine) extends FeatureWriter[SimpleFeatureType, SimpleFeature] {
+
+    private val tempFeatureIds = new AtomicLong(0)
+    private val currentFeature: ScalaSimpleFeature = new ScalaSimpleFeature(engine.sft, "")
+    private var queued = false
+
+    override def getFeatureType: SimpleFeatureType = engine.sft
+
+    override def hasNext: Boolean = false
+
+    override def next(): SimpleFeature = {
+      var i = 0
+      while (i < engine.sft.getAttributeCount) {
+        currentFeature.setAttributeNoConvert(i, null)
+        i += 1
+      }
+      currentFeature.getUserData.clear()
+      currentFeature.setId(tempFeatureIds.getAndIncrement().toString)
+      queued = true
+      currentFeature
+    }
+
+    override def write(): Unit = {
+      if (!queued) {
+        throw new IllegalStateException("Must call 'next' before 'write'")
```

Comment:
I wonder if this is the strict contract here.  *shrugs*

Reply:
that's how we have it in our other feature writers... generally it makes more sense if you're modifying existing records but since the gt api doesn't distinguish between the two cases...